src/server/worker/api_server.go

```diff
@@ -957,3 +744,420 @@ func (a *APIServer) updateJobState(stm col.STM, jobInfo *pps.JobInfo, state pps.
 	jobs.Put(jobInfo.Job.ID, jobInfo)
 	return nil
 }
+
+type acquireDatumsFunc func(low, high int64) (failedDatumID string, _ error)
+
+func (a *APIServer) acquireDatums(ctx context.Context, jobID string, chunks *Chunks, logger *taggedLogger, process acquireDatumsFunc) error {
+	complete := false
+	for !complete {
+		var low int64
+		var high int64
+		var found bool
+		if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			found = false
+			locks := a.locks(jobID).ReadWrite(stm)
+			// we set complete to true and then unset it if we find an incomplete chunk
+			complete = true
+			for _, high = range chunks.Chunks {
+				var chunkState ChunkState
+				if err := locks.Get(fmt.Sprint(high), &chunkState); err != nil {
+					if col.IsErrNotFound(err) {
+						found = true
+					} else {
+						return err
+					}
+				}
+				// This gets triggered either if we found a chunk that wasn't
+				// complete or if we didn't find a chunk at all.
+				if chunkState.State == ChunkState_RUNNING {
+					complete = false
+				}
+				if found {
+					break
+				}
+				low = high
+			}
+			if found {
+				return locks.PutTTL(fmt.Sprint(high), &ChunkState{State: ChunkState_RUNNING}, ttl)
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+		if found {
+			// process the datums in newRange
+			failedDatumID, err := process(low, high)
```

Comment:
(counterargument: if this worker is stuck for some reason, but its goroutine keeps updating the lease indefinitely, then the job will never finish)

Reply:
Good catch, we absolutely need this, otherwise we can't process datums that take longer than 30 seconds.

Reply:
I think we basically have to resign ourselves to the fact that if a worker gets stuck because the user code never completes then the job is going to get stuck. This doesn't seem to me like a great place to insulate ourselves from that.

Reply:
Also #2503 will help a lot with this.